[SILGen]: emit default arguments after all other 'delayed' arguments

jamieQ opened the pull request:
Currently, it is impossible to make a call to a function which:

1. Takes an `inout` generic parameter
2. Contains a default argument preceding said `inout` parameter
3. Implicitly opens an existential to bind to the `inout` parameter

Should you think this is an esoteric combination, consider this attempt to print to an existential output stream:

```swift
var stream: any TextOutputStream = ""
print("hello", to: &stream) // Crashes the compiler
```

This change extends the logic implemented to support [SE-0411](https://swiftlang.github.io/swift-evolution/#?proposal=SE-0411) (isolated default arguments) to have SILGen emit default arguments only after all other forms of 'delayed arguments' have been emitted – in particular, default arguments will now be emitted after all formal argument accesses have began. This ensures the use of opened existentials within the default argument generators no longer violate SIL verifier expectations or other invariants which currently cause such code to crash.

See [additional forum discussion](https://forums.swift.org/t/thoughts-on-fixing-a-sil-verification-error-and-or-bad-codegen-involving-default-arguments-implicitly-opened-existentials/82409/4).

Resolves: https://github.com/swiftlang/swift/issues/65091

jamieQ commented:
@swift-ci Please smoke test macOS platform

slavapestov commented:
Are there "non-philosophical" reasons that we might not want to delay default arguments until after formal accesses begin?

As you said, it seems that the exclusivity violations will only occur in pathological situations, whereas as it stands, this bug is trivial to hit with basic usages of `print()` for example.

I didn't realize this bug has so many duplicates going back years, so in my opinion that makes the current fix the right approach. Opening existentials "in order" would be nice, but I think that runs the risk of breaking existing code more than delaying default arguments further does.

jamieQ commented:
> Are there "non-philosophical" reasons that we might not want to delay default arguments until after formal accesses begin?
> 
> As you said, it seems that the exclusivity violations will only occur in pathological situations, whereas as it stands, this bug is trivial to hit with basic usages of `print()` for example.

personally i just felt less comfortable/confident with the change given its potential to break existing code at runtime vs the fact that the current issue just prevents certain constructs from compiling without some workaround boilerplate. are there 'legitimate' cases in which a formal storage access could also occur in a default argument... 🤔? also it would be observable via side effects since evaluation order would change, so there's a conservative argument that the reordering should maybe only be applied in the open-existential-lvalue case (which IIUC you don't find compelling enough to be worth the complexity).

> I didn't realize this bug has so many duplicates going back years, so in my opinion that makes the current fix the right approach. Opening existentials "in order" would be nice, but I think that runs the risk of breaking existing code more than delaying default arguments further does.

if you think this is a good path forward, happy to make another pass to clean it up and we can see how it goes.

jamieQ committed:
[SILGen]: emit default arguments after all other 'delayed' arguments

Currently, it is impossible to make a call to a function which:

    1. Takes an `inout` generic parameter
    2. Contains a default argument preceding said `inout` parameter
    3. Implicitly opens an exeistential to bind to `inout` parameter

Should you think this is an esoteric combination, consider this attempt
to print to an existential output stream:

```swift
var stream: any TextOutputStream = ""
print("hello", to: &stream) // Crashes the compiler
```

This change extends the logic implemented to support SE-0411 (isolated
default arguments) to apply the additional deferral of default argument
evaluation to occur after all other forms of delayed arguments – in
particular after formal argument accessed have began.

jamieQ commented:
@swift-ci test source compatibility

jamieQ commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -3184,7 +3209,7 @@ static void emitDelayedArguments(SILGenFunction &SGF,
     }
 
-    // The index saved in isolatedArgs is the index where we're
+    // The index saved in deferredDefaultArgs is the index where we're
```
- [ ] TODO: fix

jamieQ commented:
@slavapestov @kavon – okay the general idea and implementation changes here are probably in a decently 'reviewable' state now, so i think what would be helpful to further this is:

1. a review
2. pointers on what more effective test cases would look like (currently they're a bit skeletal). frankly i'm a bit surprised nothing currently in the test suite appeared to care about shuffling all default arguments around like this.

also i would very much be interested in @hborla or @rjmccall's thoughts on this, as they seem to have been the most involved in the implementation of isolated default arguments, upon which these changes are mostly built.
